### PR TITLE
Enabling FlashAttention for SDPA when given NestedTensor

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Previously, for NestedTensor inputs flash_attention was disabled due to an Illegal Memory Access error that was occurring on the "cutlass" branch of flash-attention that had be incorporated into core.  Since we have switched to the main branch of flash_attention we the existing repro script did not produce the same memory error. This PR re-enables the FlashAttention Path for NTs. As well it unifies the nested preprocessing between the two implementations. 

cc @cpuhrsch @jbschlosser @bhosmer @mikaylagawarecki

### CODE REVIEW

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 1]
```diff
@@ -474,84 +420,112 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_efficient_attention_nestedtensor_
...
+  auto
+      [query_buffer_reshaped,
+       key_buffer_reshaped,
+       value_buffer_reshaped,
+       cumulative_sequence_length_q,
+       cumulative_sequence_length_k,
+       max_seqlen_batch_q,
+       max_seqlen_batch_kv,
+       output_shape] = sdpa_nested_preprocessing(query, key, value);
+  TORCH_CHECK(max_seqlen_batch_q == max_seqlen_batch_kv)
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:514]: n00b question on my end, why do we need this TORCH_CHECK?

I've been running with max_seqlen_batch_q = 1 and  max_seqlen_batch_kv = smth else and it seemed to work

btw, what does [this part](https://github.com/pytorch/pytorch/blob/325b43661e44c99fb3a95bb2a4a94ad79f315c9d/aten/src/ATen/native/transformers/cuda/attention.cu#L909-L911) where it says max_seq_len_k is set in the kernel do (is it related?)

**Review by mikaylagawarecki — APPROVED** [review]

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 2]
```diff
@@ -474,84 +420,117 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_efficient_attention_nestedtensor_
...
+
+  return std::make_tuple(
+      query_buffer_reshaped,
+      key_buffer_reshaped,
+      value_buffer_reshaped,
+      cumulative_sequence_length_q,
+      cumulative_sequence_length_kv,
+      max_seqlen_batch_q,
+      max_seqlen_batch_kv,
+      output_shape);
```
Comment (Skylion007) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:437]:
```suggestion
    return std::make_tuple(
      std::move(query_buffer_reshaped),
      std::move(key_buffer_reshaped),
      std::move(value_buffer_reshaped),
      cumulative_sequence_length_q,
      cumulative_sequence_length_kv,
      max_seqlen_batch_q,
      max_seqlen_batch_kv,
      std::move(output_shape));

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 3]
```diff
@@ -474,84 +420,117 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_efficient_attention_nestedtensor_
...
+          value_buffer_reshaped.unsqueeze(0),
+          cumulative_sequence_length_q,
+          cumulative_sequence_length_kv,
+          max_seqlen_batch_q,
+          compute_log_sumexp,
+          is_causal);
+  // Reshape output to convert nnz to batch_size and seq_len
+  Tensor attention = std::get<0>(attention_and_logsumexp);
+  attention = wrap_buffer(attention.view(-1), output_shape).transpose(1, 2);
+  return std::tie(attention, std::get<1>(attention_and_logsumexp));
```
Comment (Skylion007) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:533]:
```suggestion
  return std::tie(std::move(attention), std::get<1>(std::move(attention_and_logsumexp)));

**Comment by Skylion007** [comment]
Ah, merge already started. I'll just leave me nits here anyway.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp (hunk 2 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:437]: A few questions I have about these. If add the standard move clang-tidy will warn me with `Moving a temporary object prevents copy elision (fix available)`. I am assuming that these objects will get moved when building PyTorch with optimizations turned on.

Second question aren't Tensors trivially copyable anyways so we don't need to be moving?

**Pull request closed without merging**